packages/grid/x-data-grid-pro/src/hooks/features/rowGrouping/useGridRowGrouping.tsx

```diff
@@ -420,16 +245,31 @@ export const useGridRowGrouping = (
     GridEventListener<GridEvents.columnsChange>
   >(() => {
     const rowGroupingModel = gridRowGroupingSanitizedModelSelector(apiRef);
-    const lastGroupingColumnsModelApplied = sanitizedModelOnLastRowPreProcessing.current;
+    const lastGroupingColumnsModelApplied = gridRowGroupingStateSelector(
+      apiRef.current.state,
+    ).unstable_sanitizedModelOnLastRowTreeCreation;
 
     if (!isDeepEqual(lastGroupingColumnsModelApplied, rowGroupingModel)) {
-      sanitizedModelOnLastRowPreProcessing.current = rowGroupingModel;
+      apiRef.current.setState((state) => ({
+        ...state,
+        rowGrouping: {
+          ...state.rowGrouping,
+          unstable_sanitizedModelOnLastRowTreeCreation: rowGroupingModel,
+        },
+      }));
 
       // Refresh the column pre-processing
+      // TODO: Add a clean way to re-run a pipe processing without faking a change
```

Comment:
Here, both for the pipe and straegy processing, we ae doing a hacky thing
I have several very imperfect ideas on how to improve this later.

Reply:
A bit related to https://github.com/mui/mui-x/pull/4030#discussion_r817027799

Maybe the strategy pre-processor that should change and trigger again the pre-processing. This already happens with the pipe pre-processors where in the next render, if one processor has changed, they are called again.

Reply:
It is also the case here
But in some scenarios, you don't have any deps of the `useCallback` that changes because the change is hidden deep inside the state.

For instance, the `updateColumns` a few lines above is to trigger the `hydrateColumns` pre-process.